e2e/nomostest/testwatcher/watch.go

```diff
@@ -271,34 +271,36 @@ func (w *Watcher) WatchObject(gvk schema.GroupVersionKind, name, namespace strin
 	_, err = watchtools.Until(ctx, initialResourceVersion, lw, condition)
 	if err != nil {
 		if err == wait.ErrWaitTimeout {
-			if len(evalErrs) > 0 {
-				return &ErrWatchTimeout{
-					cause:  multierr.Combine(evalErrs...),
-					prefix: errPrefix,
-				}
+			// Until returns ErrWaitTimeout for any ctx.Done
+			// https://github.com/kubernetes/apimachinery/blob/v0.26.3/pkg/util/wait/wait.go#L594
+			// https://github.com/kubernetes/client-go/blob/v0.26.3/tools/watch/until.go#L89
+			// So use the context error instead, if done.
+			// Use errors.Is to detect DeadlineExceeded or Cancelled.
+			select {
+			case <-ctx.Done():
```

Comment:
Should we also check if `spec.timeout == 0` before switching this to the context err (i.e. the watch was configured without a timeout)?

Reply:
Until returns ErrWaitTimeout even if the context was cancelled. So using `ctx.Err()` lets us distiguish between Cancelled and DeadlineExceeded. If there's no timeout configured, it can still be cancelled.